tests/unit/test_sysctl.py

```diff
@@ -0,0 +1,332 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import unittest
+from pathlib import Path
+from subprocess import CalledProcessError
+from unittest.mock import call, mock_open, patch
+
+from charms.operator_libs_linux.v0 import sysctl
+
+permission_failure_output = """sysctl: permission denied on key "vm.swappiness", ignoring"""
+partial_permission_failure_output = """sysctl: permission denied on key "vm.swappiness", ignoring
+net.ipv4.tcp_max_syn_backlog = 4096
+"""
+
+TEST_OTHER_CHARM_FILE = """vm.swappiness=60
+net.ipv4.tcp_max_syn_backlog=4096
+"""
+TEST_MERGED_FILE = """# This config file was produced by sysctl lib v0.1
+#
+# This file represents the output of the sysctl lib, which can combine multiple
+# configurations into a single file like.
+vm.max_map_count = 262144
+vm.swappiness=0
+
+"""
+TEST_MULTIPLE_MERGED_FILE = [
+    "# This config file was produced by sysctl lib v0.2\n#\n# This file represents the output of the sysctl lib, which can combine multiple\n# configurations into a single file like.\n",
+    "vm.swappiness=60\n",
+    "net.ipv4.tcp_max_syn_backlog=4096\n",
+]
+TEST_UPDATE_MERGED_FILE = [
+    "# This config file was produced by sysctl lib v0.2\n#\n# This file represents the output of the sysctl lib, which can combine multiple\n# configurations into a single file like.\n",
+    "vm.max_map_count=25500\n",
+]
+
+
+def check_output_side_effects(*args, **kwargs):
+    if args[0] == ["sysctl", "vm.swappiness", "-n"]:
+        return "1"
+    elif args[0] == ["sysctl", "other_value", "-n"]:
+        return "5"
+    elif args[0] == ["sysctl", "vm.swappiness=1", "other_value=5"]:
+        return "1\n5"
+    elif args[0] == ["sysctl", "vm.swappiness=0"]:
+        return permission_failure_output
+    elif args[0] == ["sysctl", "vm.swappiness=0", "net.ipv4.tcp_max_syn_backlog=4096"]:
+        return partial_permission_failure_output
+    elif args[0] == ["sysctl", "exception"]:
+        raise CalledProcessError(returncode=1, cmd=args[0], output="error on command")
+
+    # Tests on 'update()'
+    elif args[0] == ["sysctl", "vm.max_map_count", "-n"]:
+        return "25000"
+    elif args[0] == ["sysctl", "vm.max_map_count=25500"]:
+        return "25500"
+
+
+class TestSysctlConfig(unittest.TestCase):
+    def setUp(self) -> None:
```

Comment:
I really think that we should mocked `SYSCTL_DIRECTORY` (probably also `SYSCTL_FILENAME`) to ensure that this tests will not create such file or change it. Right now I did not spotted any place where we are not mocking load or write config, but this can easily happen in future and we should avoid touching system configuration in unit tests.